knitr/planetary_motion/planetary_motion.rmd

```diff
@@ -0,0 +1,735 @@
+At $t = 0$, $q_0 = (1, 0)$ and $p_0 = (0, 1)$,
+and we set $k = 1$.
+```{r, fig.width=4, fig.height=3}
+mod <- cmdstan_model("model/planetary_motion_sim.stan")
+
+n <- 40
+sim <- mod$sample(data = list(n = n, sigma_x = 0.01, sigma_y = 0.01), 
+                  chains = 1, iter_warmup = 1, 
+                  iter_sampling = 2, seed = 123)
+
+r_sim <- read_stan_csv(sim$output_files())
+simulation <- rstan::extract(r_sim, pars = "q_obs")$q_obs[1, , ]
+
+q_x <- simulation[, 1]
+q_y <- simulation[, 2]
+
+q_obs <- array(NA, c(n, 2))
+q_obs[, 1] <- q_x
+q_obs[, 2] <- q_y
+
+sub_set <- seq(from = 1, to = 40, by = 2)
+plot <- ggplot(data = data.frame(q_x = q_x[sub_set], 
+                                 q_y = q_y[sub_set], 
+                                 time = sub_set), 
+               aes(x = q_x, y = q_y, label = time)) +  theme_bw() +
+  geom_text()
+plot
+```
+
+# Fitting a simple model and diagnosing inference
+
+A first attempt at fitting the complete model fails spectacularly: the chains do not converge and take a long time to run.
+This is an invitation to start with a simpler model.
+A useful simplification is more manageable but still exhibits the challenges we encounter with the complete model.
+The hope is that a fix in this simple context translates into a fix in a more sophisticated setting.
+This turns out to be quite true here where a simple one-parameter model allows us to understand the fundamentally multimodal nature of the model.
+
+There are many ways to simplify a model.
+A general approach is to fix some of the parameters, which we can easily do when working with simulated data.
+We fix all the parameters, except for $k$.
+We now want to characterize the posterior distribution,
+$$
+   p(k \mid q_\mathrm{obs}),
+$$
+by fitting the model with Stan.
+Since $k$ is a scalar, we could use quadrature but our goal is to understand the pathologies that frustrate our inference algorithm, so we stick to MCMC.
+As a prior, we use
+$$
+  k \sim \mathrm{Normal}^+(0, 1),
+$$
+which is fairly uninformative.
+
+We fit 8 chains in parallel.
+```{r }
+chains <- 8
+```
+
+```
+fit <- mod$sample(data = list(n = n, q_obs = q_obs),
+                  chains = chains, parallel_chains = chains,
+                  iter_warmup = 500,
+                  iter_sampling = 500,
+                  seed = 123, save_warmup = TRUE)
+```
+
+```{r }
+# The model takes a while to run, so we read in the saved output.
+r_fit1 <- readRDS("saved_fit/fit1.RDS")
+get_elapsed_time(r_fit1)
+```
+
+The first notable pathology is that some of the chains take a much longer time to run. The difference is not subtle...
+
+Let's examine the summary.
+```{r }
+summary(r_fit1, pars = c("lp__", "k"), probs = c())[1]
+```
+$\hat R \gg 1$. 
+Wow, these numbers are dramatic!
+Clearly the chains are not mixing and we can visualize this using trace plots.
+```{r fig.height=3}
+traceplot(r_fit1, pars = c("lp__", "k"))
```

Comment:
`bayesplot::mcmc_trace(fit, pars = c("lp__", "k"))`

Reply:
This doesn't quite work. I think you first need to extract the samples, but you can't directly pass a fit object to `mcmc_trace`. @jgabry is there a trace plot function for cmdstanr? I couldn't find one in the vignettes.

Reply:
I will guess again1 weeeee: `bayesplot::mcmc_trace(fit$draws(c("lp__", "k")))`

Reply:
yup, that works. I need to tweak the color scheme... the shades of blue are not very readable with 8 chains.

Reply:
The other thing I'm not sure how to do is add the shade over the warmup area. I have some weird hacks, but I'm not quite getting the result I want. Removing the rstan dependency is worthwhile, but frankly a massive pain. I don't think rstan is better than cmstanr, but as far as diagnostic tools go, the two packages are fairly inconsistent.

Reply:
Mm, fair enough, well I think it's fine to have a file that runs the models with cmdstan and makes csvs, and then then file uses `read_stan_csv` to pull them in and work with them like rstan objects.

You should probably pretend like you ran the commands here with rstan (so the commented out example code should be rstan).

Actually can the other script run as rstan too? And then save the files as csvs or something using posterior? That way it's rstan all around.

Reply:
Also things should not be a massive pain, so if you hit one make Jonah issues so he can fix em' (cuz I'm sure he wants to know about them tooooo)

Reply:
I prefer `cmdstanr`, because it is more up to date. I also think it's good to have case studies without the rstan dependency, in order to demonstrate features from new packages. Doing the conversion is annoying, but if it takes me one day to do it properly, than I'm ok with this.

Reply:
> but if it takes me one day to do it properly

Jonah will appreciate the feedback from the try, at least. And if we go hybrid that's not really a problem. Just a weird artifact for the repo (the reader would see all rstan stuff).